### PR TITLE
problem fixed

### PR DESCRIPTION
**Pull request by zhechao**
* several function names and value names changed
* remove update() in black hole and loop hole, all code has been moved to render()
* loop hole collision hander added, blackhole coliision handler modified
* sprites for black hole and loop hole added
* black hole gravity decreased
* collision between asteroids and black hole has been changed back, but make sure that the velocity of asteroids will be larger than 0.8/ms
* remove loop hole fade interval, which can make fading process more smooth
* the minimal distance between loop hole entry and exit added

### CODE REVIEW

**Review by RealJohnSmith — CHANGES_REQUESTED** [review]
*  There is a bug in your code. When you change the size of loop hole or black hole (also, you should add parameter to constructor, which would allow for that) the rendering breaks. Find it and fix it.

* You remember what I said about the code which chops out parts of the texture and how I want it to be different? You didn't do anything at all related to this matter. Find it in previous comments if you want to and fix it.

**Comment by zhechao** [comment]
*What do you mean about the size of loop hole and black hole?  Do you mean the Radius of black hole and loop hole? Or the size of the image which will be displayed on the map?

**Comment by RealJohnSmith** [comment]
Yes

**Comment by zhechao** [comment]
I added the blackholeRadius which does not have render and update but should have the same position as black hole does. So, black hole won't need to handle the gravity which is out of its sprite. And then, the size of black hole should be able to change with the different value of radiusScale.

**Review by RealJohnSmith — CHANGES_REQUESTED** [review]
What the hell just happened?
Best option is to revert this last commit and start again. It was so close to being approved. Now it is miles away

**Comment by zhechao** [comment]
what is wrong there? when I test it, it looks well

**Comment by zhechao** [comment]
I will change it back, and remove the black hole radius, only change the render radius

**Comment by zhechao** [comment]
Black Hole Radius delete, and most things are same except a few names of value. And also, it should be able to change the size with the change of the radius scale.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: I think it might get lost in the Discord channel, but `EncounterObject`  no longer has this method. It has been refactored to `collidesWith(other)`. Please, rewrite it so correct method is referenced.

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: Why not use just `6.0 / 10.0;`?

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: I am not sure, whether I understand John's objections correctly, but I think he means, that you should not use the constant `BH_SPRITE_RADIUS` for both cutting actual image out of sprite sheet and for rendering. Someone might come around wanting to make Blackhole bit bigger, but by increasing this constant the image extraction would be broken.

Could you please make two separate constants? One for extraction of the image from the sprite sheet and other for rendering it on the screen (since they might differ)? Also, it might be nice for us to be able to set the rendering size of black hole in the constructor (which would turn one of the constants into instance variable) ;)

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: I think I have mentioned it a few times already, but starting with this line and above your indentation is off. Please, look into it, fix it and check in Github UI, that it is ok.

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: Same as in `blackhole.js` this should be named `collidesWith(other)`.

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: Once again, please consider using different constants for rendering and image extraction.

**Review by ksu-shanemat — CHANGES_REQUESTED** [review]
I have found few places, where you can improve your code. Most of them are not critical, but few of them are. Could you please look into them? :)

File: src/space/objects/realization/blackhole.js
Comment (RealJohnSmith) [reply src/space/objects/realization/blackhole.js]: or just 0.6, since you are using decimal numbers anyway

**Comment by zhechao** [comment]
I have separated the const values for cutting image and render. 
And indentation problem should be also fixed.

**Comment by ksu-shanemat** [comment]
> I have separated the const values for cutting image and render.
> And indentation problem should be also fixed.

Yup, I think all my objections are fixed 👍. Let's wait if John still has some objections, but if not, I think you can restructure your branch one last time to have one (or two) commits and then I will approve :)

**Comment by zhechao** [comment]
If you have time, would you mind tell me how to squash my commits?
Last time, I use 'git rebase -i HEAD~[number of my commits]', and I changed 'pick' to 'squash' for every commit except the first one.
But what it did was that one more commit added into my branch rather than squash all my commits into one commit.
Should I force the push to my branch once I squash my commits locally, like 'git push origin space-fight-zhechao --force' ?

**Comment by ksu-shanemat** [comment]
> If you have time, would you mind tell me how to squash my commits?
> Should I force the push to my branch once I squash my commits locally, like 'git push origin space-fight-zhechao --force' ?

Yes, if you do not force those changes in, Git will keep on complaining, that those changes cannot be pushed. However, double check what and where are you force pushing, because if you force push something bad, me and John will get **very very mad** ;)

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: Do you need this the actual _radius_ somewhere anymore? I looked through the code and I think it might be turned to `BH_SPRITE_SIZE` or better to `BH_SPRITE_WIDTH` and `BH_SPRITE_HEIGHT`. Those two constants might then get utilized in the image extraction from the sprite sheet to finally satisfy John's objections.

I will mark the places, which should be changed with proper code:

Here it should be something as:

```
const BH_SPRITE_WIDTH = 100;
const BH_SPRITE_HEIGHT = 100;

const BH_SPRITE_SHEET_WIDTH = 6;
const BH_SPRITE_SHEET_HEIGHT = 2;
```

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js:61]: Wrap lines 60-74 to a separate function called something like `updateImageSource(deltaT)` and replace the hard-coded constants with appropriate constants set in the top of the file e.g. as follows:

```
this.wrapInterval -= deltaT;
if(this.wrapInterval <= 0) {
  if(this.coordinateXInImage == (BH_SPRITE_SHEET_WIDTH - 1) * BH_SPRITE_WIDTH) {
     this.coordinateXInImage = 0;
     if(this.coordinateYInImage == (BH_SPRITE_SHEET_HEIGHT - 1) * BH_SPRITE_HIEGHT) {
        this.coordinateYInImage = 0;
      } else {
        this.coordinateYInImage += BH_SPRITE_HIEGHT;
      }
  } else {
     this.coordinateXInImage += BH_SPRITE_WIDTH;
  }

  this.wrapInterval = BH_WRAP_INTERVAL;
}
```

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/blackhole.js]: This then can be also refactored to:

```
context.drawImage( this.bhImage, this.coordinateXInImage, this.coordinateYInImage,
BS_SPRITE_WIDTH, BH_SPRITE_HEIGHT, this.position.x - renderRadius, this.position.y - renderRadius, 
2*renderRadius, 2*renderRadius );
```

Also please consider adding some line breaks to this extremely long line ;)

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js:39]: Please inspect code of this method in Github UI ... Do you really think your indentation is ok? o.O

image

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: I would also prefer using `LH_SPRITE_WIDTH` and `LH_SPRITE_HEIGHT` but it is not as crucial as in BlackHole.

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: Once again, replace those strange numbers with constants, please.

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [inline src/space/objects/realization/loop-hole.js]: Refactor those two lines according to the changes proposed in BlackHole.

**Review by ksu-shanemat — CHANGES_REQUESTED** [review]
I have talked with John and I agree, that your code could use some constants instead of seemingly random numbers. Because of the nearing deadline, I have attached proposed changes :)

File: src/space/objects/realization/loop-hole.js
Comment (zhechao) [reply src/space/objects/realization/loop-hole.js]: Can I use LH_SPRITE_SIZE, since those two values should be the same

File: src/space/objects/realization/blackhole.js
Comment (ksu-shanemat) [reply src/space/objects/realization/blackhole.js:61]: I have just realized that this code can be even simplified:

```
this.wrapInterval -= deltaT;
if(this.wrapInterval <= 0) {
  this.coordinateXInImage += BH_SPRITE_WIDTH;
  
  if(this.coordinateXInImage == BH_SPRITE_SHEET_WIDTH * BH_SPRITE_WIDTH) {
     this.coordinateXInImage = 0;
     this.coordinateYInImage += BH_SPRITE_HIEGHT;
 
     if(this.coordinateYInImage == BH_SPRITE_SHEET_HEIGHT * BH_SPRITE_HIEGHT) {
        this.coordinateYInImage = 0;
     }
  }

  this.wrapInterval = BH_WRAP_INTERVAL;
}
```

This code might not be fully functional since I am assembling it in Github UI, but will hopefully give you the idea, what has John aimed for all the time ;)

File: src/space/objects/realization/blackhole.js
Comment (zhechao) [reply src/space/objects/realization/blackhole.js]: I used BH_SPRITE_SIZE for width and height

Comment (ksu-shanemat) [reply src/space/objects/realization/blackhole.js]: Ok, this takes away some flexibility, but it is certainly ok :)

File: src/space/objects/realization/loop-hole.js
Comment (ksu-shanemat) [reply src/space/objects/realization/loop-hole.js]: Same as before, it is ok provided no one will ever decide to have rectangular loopholes :)

**Comment by zhechao** [comment]
all conversations resolved, and all commits have been squashed to one commit

**Review by ksu-shanemat — APPROVED** [review]

**Comment by zhechao** [comment]
Should I wait for John's approval?
And also, Should I merge by myself or it will auto-merge into space-master after John approved?

**Review by RealJohnSmith — APPROVED** [review]

**Comment by RealJohnSmith** [comment]
> Should I wait for John's approval?
> And also, Should I merge by myself or it will auto-merge into space-master after John approved?

It will not auto merge, someone has to. You should never merge it yourself (even if you have approvals), usually the last one who is doing the approve will merge it (in this case me)

**Pull request merged**
